src/Features/Lsif/Generator/Generator.cs

```diff
@@ -335,6 +365,31 @@ void MarkImplementationOfSymbol(ISymbol baseMember)
             return documentVertex.GetId();
         }
 
+        private static async Task GenerateSemanticTokensAsync(
+            Document document,
+            ILsifJsonWriter lsifJsonWriter,
+            IdFactory idFactory,
+            SourceText text,
+            LsifDocument documentVertex)
+        {
+            // Compute colorization data.
+            var data = await SemanticTokensHelpers.ComputeSemanticTokensDataAsync(
+                document,
+                SemanticTokensHelpers.TokenTypeToIndex,
+                new LspProtocol.Range()
+                {
+                    Start = new Position(0, 0),
+                    End = new Position(text.Lines.Count - 1, text.Lines[text.Lines.Count - 1].EndIncludingLineBreak)
+                },
+                Classification.ClassificationOptions.Default,
+                includeSyntacticClassifications: true,
```

Comment:
@allisonchou @dibarbet this flag worries me.  in what scenarios do we not want our tree-classifications as part of semantic-tokens (which i thought existed to get the real language classification tokens for all positions in the file).

Reply:
> in what scenarios do we not want our tree-classifications as part of semantic-tokens

I added a comment to clarify. Even in LSP scenarios, your accurate syntax classification is in use, it just comes through the in-proc tagger still instead of LSP classification.

- The general guidance for LSP languages is to use 'TextMate' for tokenization and LSP just for semantic classification. This is somewhat arbitrary though and based almost entirely on a convention that arose from evolutionary history of VS Code which supported TextMate long before semantic tokens and which runs TextMate _synchronously_.

- Roslyn is the one exception that uses a synchronous tagger instead of TextMate. The benefits of this tagger are that it can be synchronous (VS Windows + VS Mac TextMate is _async_ and 'chases the cursor' when typing). Your tagger is likely much more accurate than TextMate.

- There's no reason you couldn't use semantic tokens for everything (Razor does), it's just async and currently has that 'chasing the cursor' behavior, so until that's fixed, the experience might feel a bit laggy to colorize in languages that rely solely on LSP.

Reply:
> There's no reason you couldn't use semantic tokens for everything (Razor does), it's just async and currently has that 'chasing the cursor' behavior, so until that's fixed, the experience might feel a bit laggy to colorize in languages that rely solely on LSP.

Interesting/odd.  Are we not SpanTrackingMode.EdgeInclusive on these regions?  I would expect them to track with the edit that way.

> The general guidance for LSP languages is to use 'TextMate' for tokenization and LSP just for semantic classification. This is somewhat arbitrary though and based almost entirely on a convention that arose from evolutionary history of VS Code which supported TextMate long before semantic tokens and which runs TextMate synchronously.

yeah, i feel like we're in a weird middle ground.  There seems to be the recognition that synchronous classification has virtues for performance/experience.  But the solution is to use a really bad system that can't handle complex languages well.  I wish there was some way to get the best of both worlds here (like we have in VS).

Reply:
> Interesting/odd. Are we not SpanTrackingMode.EdgeInclusive on these regions? I would expect them to track with the edit that way.

We use SpanTrackingMode.EdgeExclusive. This was an intentional change because of the supported modes, it seems to misbehave in the fewest scenarios. While 'growing' the colors as the user types works well for class definitions, XML tags, etc. it causes weird effects with tokens like keywords, highlighted punctuation, etc.

A more robust option might be to duplicate what VS Code does and have a custom span tracking mode that uses a regex to decide the bounds of an identifier and uses EdgeInclusive or EdgeExclusive based on that.

> I wish there was some way to get the best of both worlds here (like we have in VS).

I'm in favor of fixing the paper cuts required to make LSP work for all tokens.

Reply:
@gundermanc does adding syntactic colorization significantly improve the experience? I know in your original spec we had landed on only doing semantic

Reply:
Ahhh love the explanation thank you! I agree with your thoughts here